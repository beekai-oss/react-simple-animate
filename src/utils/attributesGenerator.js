// @flow
import type { Props } from '../animate';
import mapSequenceOverProps from './mapSequenceOverProps';

export default function attributesGenerator(props: Props, willComplete: boolean = false, isMountWithPlay: mixed = false): Object {
  const { animationStates, sequenceId, sequenceIndex } = props;
  const id = sequenceId || sequenceIndex;
  const {
    play,
    startStyle,
    endStyle,
    onCompleteStyle,
    durationSeconds = 0.3,
    delaySeconds = 0,
    easeType = 'linear',
    className,
    reverseDurationSeconds = 0,
    reverseDelaySeconds = 0,
    forwardedRef,
  } = mapSequenceOverProps(props, id);
  let style = startStyle;
  let transition = `all ${durationSeconds}s ${easeType} ${delaySeconds}s`;

  if (!play && (reverseDurationSeconds || reverseDelaySeconds)) {
    transition = `all ${reverseDurationSeconds || durationSeconds}s ${easeType} ${reverseDelaySeconds}s`;
  } else if (!isMountWithPlay) {
    if (willComplete && onCompleteStyle && play) {
      style = onCompleteStyle;
      transition = '';
    } else if (
      play ||
      ((id || id === 0) &&
        animationStates &&
        Object.keys(animationStates).length &&
        animationStates[id] &&
        animationStates[id].play)
    ) {
      style = endStyle;
    }
  }

  return {
    className,
    style: {
      ...style,
      transition,
    },
    ref: forwardedRef,
  };
}
