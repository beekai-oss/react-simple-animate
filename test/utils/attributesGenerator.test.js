import attributesGenerator from '../../src/utils/attributesGenerator';

jest.mock('../../src/utils/mapSequenceOverProps', () => props => props);

describe('attributesGenerator', () => {
  const state = {
    willComplete: false,
  };

  const props = {
    startStyle: {
      display: 'none',
    },
    endStyle: {
      display: 'block',
    },
    easeType: 'ease',
    play: false,
  };

  const startStyleWithTransition = {
    ...props.startStyle,
    transition: 'all 0.3s ease 0s',
  };

  const endStyleWithTransition = {
    ...props.endStyle,
    transition: 'all 0.3s ease 0s',
  };

  it('should generate startStyle', () => {
    expect(attributesGenerator(props, state).style).toEqual(startStyleWithTransition);
  });

  describe('when animation is set to true', () => {
    it('should generate endStyle', () => {
      expect(attributesGenerator({ ...props, play: true }, state).style).toEqual(endStyleWithTransition);
    });
  });

  describe('when animation reverseDelaySeconds has been set', () => {
    it('should render the end style', () => {
      expect(attributesGenerator({ ...props, play: false, reverseDelaySeconds: 0.3 }, state).style).toEqual({
        ...startStyleWithTransition,
        transition: 'all 0.3s ease 0.3s',
      });
    });
  });

  it('should render start style when delay second is set and animation about to start', () => {
    expect(attributesGenerator({ ...props, play: true, delaySeconds: 0.3 }, state).style).toEqual({
      ...endStyleWithTransition,
      transition: 'all 0.3s ease 0.3s',
    });
  });

  it('should apply onComplete style', () => {
    expect(
      attributesGenerator(
        { ...props, play: true, delaySeconds: 0.3, onCompleteStyle: { display: 'inline' } },
        { willComplete: true },
      ).style,
    ).toEqual({
      display: 'inline',
      transition: '',
    });
  });

  it('should apply end style when animationStates is set with play', () => {
    expect(
      attributesGenerator(
        {
          ...props,
          animationStates: { test: { play: true } },
          sequenceId: 'test',
        },
        state,
      ).style,
    ).toEqual(endStyleWithTransition);

    expect(
      attributesGenerator(
        {
          ...props,
          animationStates: { 1: { play: true } },
          sequenceIndex: 1,
        },
        state,
      ).style,
    ).toEqual(endStyleWithTransition);
  });

  it('should set up refCallback if it is part of the prop', () => {
    const refCallback = jest.fn();

    expect(
      attributesGenerator(
        {
          refCallback,
        },
        state,
      ),
    ).toMatchSnapshot();
  });
});
