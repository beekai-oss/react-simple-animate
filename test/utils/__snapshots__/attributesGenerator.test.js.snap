// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributesGenerator should set up refCallback if it is part of the prop 1`] = `
Object {
  "className": undefined,
  "ref": undefined,
  "style": Object {
    "transition": "all 0.3s linear 0s",
  },
}
`;
